site/source/tutorials/1015_form_validation/index.md

```diff
@@ -0,0 +1,366 @@
+	errorMessage?: string;
+	onValidate?: (event: Event) => void;
+}
+
+export const ValidatedTextInputBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class ValidatedTextInput extends ValidatedTextInputBase<ValidatedTextInputProperties> {
+	private _errorId = uuid();
+
+	protected render() {
+		const {
+			disabled,
+			label,
+			maxLength,
+			minLength,
+			name,
+			placeholder,
+			readOnly,
+			required,
+			type = 'text',
+			value,
+			invalid,
+			errorMessage,
+			onBlur,
+			onInput
+		} = this.properties;
+
+		return v('div', { classes: this.classes(css.inputWrapper) }, [
+			w(TextInput, {
+				describedBy: this._errorId,
+				disabled,
+				invalid,
+				label,
+				maxLength,
+				minLength,
+				name,
+				placeholder,
+				readOnly,
+				required,
+				type,
+				value,
+				onBlur,
+				onInput
+			}),
+			invalid === true ? v('span', {
+				id: this._errorId,
+				classes: this.classes(css.error),
+				'aria-live': 'polite'
+			}, [ errorMessage ]) : null
+		]);
+	}
+}
+```
+{% endsolution %}
+
+{% instruction 'Use `ValidatedTextInput` within `WorkerForm`' %}
+
+Now that `ValidatedTextInput` exists, let's swap it with `TextInput` in `WorkerForm`, and write some error message text while we're at it:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:75-115 %}
+
+{% task 'Create `onFormValidate` method separate from `onFormInput`' %}
+
+{% instruction 'Update the context to pass in an `onFormValidate` method' %}
+
+Currently the validation logic is unceremoniously dumped in `formInput` within `ApplicationContext.ts`. Now let's break that out into its own `formValidate` function, and borrow the `onFormInput` pattern to pass `onFormValidate` to `WorkerForm`. There are three steps to this:
+
+1. Add a `formValidate` method to `ApplicationContext.ts` and update `_formErrors` there instead of in `formInput`:
+	{% include_codefile 'demo/finished/biz-e-corp/src/ApplicationContext.ts' lines:71-79 %}
+2. Update `WorkerFormContainer` to pass `formValidate` as `onFormValidate`:
+	{% include_codefile 'demo/finished/biz-e-corp/src/containers/WorkerFormContainer.ts' lines:6-10 %}
+3. Within `WorkerForm`, create internal methods for each form field's validation using `onFormValidate`, and pass those methods (e.g. `onFirstNameValidate`) to each `ValidatedTextInput` widget. This should follow the same pattern as `onFormInput` and `onFirstNameInput`, `onLastNameInput`, and `onEmailInput`:
+	{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:52-62 %}
+
+{% instruction 'Handle calling `onValidate` within `ValidatedTextInput`' %}
+
+You might have noticed that the form no longer validates on user input events. This is because we no longer handle validation within `formInput` in `ApplicationContext.ts`, but we also haven't added it anywhere else. To do that, add the following private method to `ValidatedTextInput`, and pass `onInput: this._onInput` to `TextInput`:
```

Comment:
probably need to explicitly show the change to the properties that uses `this._onInput` over just `onInput` from the properties.

Reply:
There isn't an easy way to highlight single lines with a code block, is there? There have been a few places where I really wish I could 😕

Reply:
you could just do the one line that has changed maybe?

```
{% include_codefile 'demo/finished/biz-e-corp/src/file.ts' line:14 %}
```

Reply:
I suppose, but in this case it doesn't give much context. It would just be
`onInput: this._onInput`
whereas seeing it within the `w(TextInput, { ... })` makes it more obvious what should happen

Reply:
Yeah being able indicate which it the changes code in context would be much better